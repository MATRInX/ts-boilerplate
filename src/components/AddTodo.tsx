import React, { FormEvent } from 'react';
import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { addTodo } from '../actions/todos';
import { AddTodoAction } from '../actions/todos';
import { Link } from 'react-router-dom';

type AddTodoStateProps = null;

interface AddTodoDispatchProps {
  handleSubmit: (name: string) => void
}

interface AddTodoExtraProps {
  defaultValue?: string
}

type AddTodoProps = AddTodoStateProps & AddTodoDispatchProps & AddTodoExtraProps;

export interface AddTodoState {
  value: string
};

export class AddTodo extends React.Component<AddTodoProps, AddTodoState>{
  constructor(props: AddTodoProps){
    super(props);
    this.state = { value: this.props.defaultValue ? this.props.defaultValue : ''};
  }

  onChange = (e: FormEvent<EventTarget>):void => {
    e.preventDefault();
    const target = e.target as HTMLInputElement;
    const name: string = target.value;
    this.setState({ value: name });
  }

  onSubmit = (e: FormEvent<EventTarget>): void => {
    e.preventDefault();
    this.props.handleSubmit(this.state.value);
    this.setState({ value: ''});
  }

  render() {
    return (
      <div>
        <Link to="/"><h1>Todo list!</h1></Link>
        <form onSubmit={this.onSubmit}>
          <input type="text" value={this.state.value} onChange={this.onChange}/>
          <button type="submit">Add todo!</button>
        </form>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch: Dispatch<AddTodoAction> , ownProps: AddTodoProps): AddTodoDispatchProps =>  ({
  handleSubmit: (name: string) => dispatch(addTodo(name))
})

//export default connect<AddTodoStateProps, AddTodoDispatchProps, AddTodoProps>
export default connect<AddTodoStateProps, AddTodoDispatchProps, any>
  (null, mapDispatchToProps)(AddTodo);