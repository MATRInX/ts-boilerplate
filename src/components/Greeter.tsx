import React, { SFC } from 'react';

export interface GreeterProps {
  name?: string
};

export const Greeter: SFC<GreeterProps> = (props) => {
  const { name } = props;
  return (
    <div>
      Hello {name} in world of typescript!
    </div>
  );
}

export default Greeter;