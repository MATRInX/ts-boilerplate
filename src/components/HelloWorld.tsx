import React, { SFC } from 'react';

export const HelloWorld: SFC<{}> = () => (
  <div>
    Hello World from TS!
  </div>
);

export default HelloWorld;