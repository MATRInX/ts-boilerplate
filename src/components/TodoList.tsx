import React from 'react';
import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import Todo from '../models/Todo';
import * as fromTodo from '../reducers/todos';
import { toggleTodo, ToggleTodoAction } from '../actions/todos';
import { Link } from 'react-router-dom';

export interface TodoListStoreProps {
  todos: fromTodo.State
}

export interface TodoListDispatchProps {
  onTodoClick: (todoId: number) => void
}

type TodoListProps = TodoListStoreProps & TodoListDispatchProps;

interface TodoListState { }

export class TodoList extends React.Component<TodoListProps, TodoListState> {
constructor(props: TodoListProps){
  super(props);
}

  onTodoClick = (todoId: number): void => {
    this.props.onTodoClick(todoId);
  }

  render() {
    const { todos, onTodoClick } = this.props;
    return (
      <div>
        <Link to="/add"><h2>Add todos...</h2></Link>
        {
          this.props.todos.length > 0 ? (
            this.props.todos.map((todoItem: Todo) => 
            <div key={todoItem.id} onClick={() => this.onTodoClick(todoItem.id)}>
              {todoItem.name} <span>{todoItem.done ? 'is DONE' : 'has to be done!'}</span>
            </div>)
          ) : (
            <p>There is no items!</p>
          )
        }
      </div>
    )
  }
}

const mapStateToProps = (storeState: TodoListStoreProps, ownProps: TodoListProps) => ({ 
  todos: storeState.todos
});

const mapDispatchToProps = (dispatch: Dispatch<ToggleTodoAction>, ownProps: TodoListProps) => ({
  onTodoClick: (todoId: number) => dispatch(toggleTodo(todoId))
});

export default connect<TodoListStoreProps, TodoListDispatchProps, any, any>
  (mapStateToProps, mapDispatchToProps)(TodoList);