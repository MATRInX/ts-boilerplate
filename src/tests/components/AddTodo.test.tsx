import React from 'react';
import { shallow } from 'enzyme';
import { AddTodo } from '../../components/AddTodo';
// import { wrap } from 'module';

let handleSubmit: jest.Mock;
let wrapper: any;
const defaultValue: string = 'Default value';

beforeEach(() => {
  handleSubmit = jest.fn();
  wrapper = shallow(<AddTodo handleSubmit={handleSubmit} defaultValue={defaultValue}/>);
})

test('should correctly render AddTodo component', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should set todo text on input change', () => {
  const value: string = 'Sample todo item.';
  expect(wrapper.state('value')).toBe(defaultValue);
  wrapper.find('input').simulate('change', { 
    target: { value },
    preventDefault: () => {} 
  });
  expect(wrapper.state('value')).toBe(value);
});

test('should submit form with valid data', () => {
  wrapper.find('form').simulate('submit', { preventDefault: () => {}});
  expect(handleSubmit).toHaveBeenCalledWith(defaultValue);
  expect(wrapper.state('value')).toBe('');
});