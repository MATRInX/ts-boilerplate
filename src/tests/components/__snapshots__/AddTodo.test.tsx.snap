// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render AddTodo component 1`] = `
<div>
  <Link
    replace={false}
    to="/"
  >
    <h1>
      Todo list!
    </h1>
  </Link>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
      value="Default value"
    />
    <button
      type="submit"
    >
      Add todo!
    </button>
  </form>
</div>
`;
